### ISSUE TITLE
Selecting features of this particular shapefile results in different features being highlighted than the ones expected

### ISSUE THREAD

**Issue by VectorZita** [DotSpatial/DotSpatial#1134]
Hello to all, I am new to dot spatial and I am trying to learn and appreciate the most of this remarkable application (I apologize in advance for not being word-perfect in English, I am still learning!).

I am facing something a bit peculiar: I open [this shapefile](http://ais-faa.opendata.arcgis.com/datasets/2f392b8215304e85bc835addafe2ca42_0.zip).

When I select a feature(s), random ones are drawn highlighted instead (though the correct ones are selected).
This happens also when I open this shapefile in demomap.
My guess is that something is wrong with the particular shapefile, since none of the others I use shows this kind of behavior, but any ideas what might cause this?

**In a nutshell:**

### Steps to reproduce
1. Download [this shapefile](http://ais-faa.opendata.arcgis.com/datasets/2f392b8215304e85bc835addafe2ca42_0.zip)
2. Open it in demomap and select a feature(s)

###Expected behavior:
Map shows highlighted the selected feature(s)

###Actual behavior:
Although I can confirm that application selects (internally) the actual features, random ones are depicted highlighted.

DotSpatial version: 2.0

Screenshots:

I try to select this feature:
https://i.imgur.com/kfow06g.png

These are actually highlighted:
https://i.imgur.com/wWnGyH0.png

Thank you in advance for your time and all the efforts you put into this.

V.Z.

**Comment by MatthiasSchilder** [comment]
Hi,

your shapefile contains 387 which do not have a geometry. You can identify them, for example, with QGIS' expression editor and a ' $geometry IS NULL'-query.

If you fix the file (delete features without geometry), it should give you the expected result (I did a rectangular select for this example):
![image](https://user-images.githubusercontent.com/313898/37107356-6bd6e36c-2234-11e8-8a75-2821b9f35fae.png)

I think there is maybe a bug corresponding to how DotSpatial checks, if a feature should be drawn as selected or not selected, in cases where features without geometry exist. I'm currently investigating this.

**Comment by MatthiasSchilder** [comment]
When the shapefile is initially read, a consecutive array containing all features coordinates is created. It's length is equal to the count of features WITH geometries in the featureset. When it comes to drawing, an second array is created which holds information about feature selection state, visibility state and which category symbolizer to use for this feature (drawnstates). The length of this array equals the count of ALL features in the feature set, not only those with a geometry.

At each draw, feature geometries are intersected with the view extents. If a feature's geometry is within the view extents, the index which points to the geometry of this feature is placed in a list of all features to draw. Later, this index is used to get it's geometry from the vertex-array (this access yields the correct geometry) and the features drawnstate from the drawn-states array. The last access yields a wrong drawn-state, since this array must be indexed with the feature-index, not the geometry-index.

I have created a proposal for a fix. It got appended to my last already open pull request. I did only address the issue for point objects, maybe it does also occur for other feature types.

**Comment by VectorZita** [comment]
I am really grateful for your thorough answer! I will find a workaround until it is fixed. Again thank you for your effort, this bug was bugging me (no pun intended) a lot!

**Comment by UliEgger** [comment]
We have an updated proposal based on the research of Matthias Schilder but which avoids another problem created by it.
